### ISSUE TITLE
How to reproduce a Cross Entropy Loss without losing numerical accuracy?

### ISSUE THREAD

**Issue by tamerlansb** [pytorch/pytorch#25990]
## 🐛 Bug

I get a discrepancy between the values of the losses obtained by the torch. CrossEntropyLoss and CustomeCrossEntropyLoss.

## To Reproduce

    import torch
    import torch.nn.modules.loss as L  

    class CustomCrossEntropyLoss(L._Loss):
        def __init__(self, reduction=True):
            super(CustomCrossEntropyLoss, self).__init__()
            self.reduction = reduction
    
         def forward(self, inp, target):
             input_target = inp.gather(1, target.view(-1, 1))
             input_max, _ = inp.max(dim=1, keepdim=True)
             output_exp = torch.exp(inp - input_max)
             output_softmax_sum = output_exp.sum(dim=1)
             output = -input_target + torch.log(output_softmax_sum).view(-1, 1) + input_max
             if self.reduction:
                 output = output.mean()
             return output

     torch_ce = torch.nn.CrossEntropyLoss(reduction='none')
     custom_ce = CustomCrossEntropyLoss(reduction=False)
     batch_size = 128
     N_class = 90000
     logits = torch.randn((batch_size,N_class))
     targets = torch.randint(N_class, (batch_size,))
     print((torch_ce(logits, targets).view(-1) - custom_ce(logits,targets).view(-1)).mean())

## Expected behavior

I get a minimal non-zero discrepancy of the order of 10e-7, which then occurs in gradients and can affect the learning outcome.
How to fix this problem?

## Environment

       Collecting environment information...
	PyTorch version: 1.2.0
	Is debug build: No
	CUDA used to build PyTorch: 10.0.130

	OS: CentOS Linux 7 (Core)
	GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
	CMake version: version 2.8.12.2

	Python version: 3.7
	Is CUDA available: Yes
	CUDA runtime version: Could not collect

	Nvidia driver version: 430.14
	cuDNN version: Could not collect

	Versions of relevant libraries:
	[pip3] numpy==1.15.4
	[pip3] tensorboard-pytorch==0.7.1
	[pip3] torch==1.0.0
	[pip3] torchvision==0.2.1
	[conda] blas                      1.0                   mkl.conda
	[conda] mkl                       2019.4                243.conda
	[conda] mkl-service               2.0.2           py37h7b6447c_0.conda
	[conda] mkl_fft                   1.0.12          py37ha843d7b_0.conda
	[conda] mkl_random                1.0.2           py37hd81dba3_0.conda
	[conda] pytorch                   1.2.0           py3.7_cuda10.0.130_cudnn7.6.2_0    pytorch
	[conda] torchvision               0.4.0                py37_cu100    pytorch

**Comment by soumith** [comment]
please use https://discuss.pytorch.org for questions
